### ISSUE TITLE
Converting Website

### ISSUE THREAD

**Issue by almic** [almic/CSGO-Skill#16]
I believe that converting this to a Node.js web app with Express will allow for much more control and much less hassle in the future. Also, it will greatly reduce costs. Here's a look at the current price table:

- Hosting: $80 ~ $140 / year
- Domain: $10 ~ $40 / year
- Total: $90 ~ $150+ / year

You might be thinking "Why are you writing this, no one is here except you?" And to that I say "I know, but this is mostly for historical reasons; I like to keep records of decisions."

I used to be paying at least $80 per year to host with GoDaddy, the cheapest (non-free) option I could manage. I also used to be paying over $30 per year to keep a domain name. Recently, I discovered that Google offers `.com` domains as cheap as $12 per year. So I greatly reduced the cost there. Very recently I tried GreenGeeks, who offered a bit higher premium but with a much more up-to-date environment. GoDaddy was stuck on a year-old cPanel version, while GreenGeeks was basically running the latest version. They also offered more space and had a crazy discount for me. I paid $35 for a year of better-than-godaddy everything, and right on time because I was set to renew hosting in a few months. This seemed to be a no-brainer.

However, things went down hill rather rapidly. First of all, they are in the stone age for security. They "required" me to set a PIN (there's a lot more detail here but I'm going to skip it). Yes, a fucking PIN. Apparently a PIN is more secure than an actual password, of which mine looks something like UHD*I#^GRI$*HF&. I just spammed my keyboard to type that, but you get the idea. They also forced me to contact support to change the domain associated with the server, which meant I had to wait over 24 hours after buying to correctly set the domain (I used a different one than I intended to test the service, silly me!) There were also some storage limitations, and you can forget about subdomains. They forced them to exist inside the public_html directory, which gave me a headache to think about. There were many other stupid things wrong with their system, and in only a month I had contacted support 3 times to ask about why something wasn't working and asking them to fix some parts of my account I managed to screw up.

It got to the point where I wish I had rather stuck with Godaddy instead of spending $35 on a Steam Support simulator. Also, I found that I would end up renewing for well over $130 at the end of my term. After contacting sales support I was able to lower it to somewhere between $150~$200 so long as I renewed for 2 years and kept the same plan. Maybe I was a bit blinded by the insane discount to notice the actual cost of what I was getting. Ironically, I fell for the same trick with GoDaddy, paying about $20 for a year of hosting, and then renewing for over $90 with added services I didn't realize I had, and never even used or asked for. So far I'm pretty sure I could've bought a GTX1080 with the amount of money I've spent trying to figure this shit out. Luckily I managed to refund some of it, but I've still spent too much just to find a good host.

About 3 months ago I discovered Heroku, which is nice enough to basically let you run one Node.js application 24/7 absolutely free, with pretty okay stats. I kinda forgot about it, though. I wrote a simple Discord bot just to dip my feet into Node.js as a fun little side project. It was amazing, I wrote my code, pushed a commit to my GitHub repo, and in a few seconds my Discord bot went offline, and came back online with the updated code, ready for me to communicate with it in real time with very little latency. My discord bot has actually been running non-stop for 3 months now, and as far as I can tell, it has not gone offline since I last touched it. That's good enough up-time for me.

For the past few days, I've been searching for a cheaper alternative to hosting, and realized that Node.js + Express had basically everything I need. Now came to the database. Heroku offers some free database options, but they are extremely limited in storage. The best option I could find had 5mb and 10 connections, and from the flavor text it sounded like I'd be lucky to have 5 active users not experience delays, let alone >10.

I found out that MongoDB themselves offer a pretty attractive free option. They give you 512mb with 100 max connections, and bonus points for being MongoDB. So basically I'm on my merry way to building a MEAN app, all in the name of cost reduction, without even realizing it. Of course there is the possibility of needing to scale up, but I don't want to think about it right now.

Anyway, here's the new figures:
- Hosting: $0/ month with Heroku
- Domain: $1/ month with Google
- MongoDB: $0/ forever with MongoDB
- Total: $12/ year

If I can really lower it down that far, then I'd want to be transparent about it. I don't want people to guess how much this project is costing, I want them to know how little or how much it is and how many strings I'm pulling on to make it work.

All I have to do is write the server, redo all the PHP into javascript, figure out how to use MongoDB, and write up the general strategy for how to do all that so I can do it correctly the first time. Anyway, lots more details I won't go into but the idea here is that I can potentially cut costs by over 700%

tl;dr I can save a lot of money by using Node.js

**Comment by almic** [comment]
Will be using issue #18 from now on
